scale/ingest/scan/scanners/scanner.py

```diff
@@ -0,0 +1,317 @@
+            invalid
+        """
+
+        raise NotImplementedError
+
+    def _callback(self, file_list):
+        """Callback for handling files identified by list_files callback
+        
+        :param file_list: List of files found within workspace
+        :type file_list: storage.brokers.broker.FileDetails
+        """
+
+        ingests = []
+
+        for file_details in file_list:
+            if not self._stop_received:
+                ingest = self._ingest_file(file_details.file, file_details.size)
+                # Only bother appending ingests 
+                if not self._dry_run:
+                    ingests.append(ingest)
+                self._count += 1
+            else:
+                raise ScannerInterruptRequested
+
+        # If no ingests were added, don't bother moving on
+        if not len(ingests):
+            logger.debug('No ingests for batch, this will always be the case during a dry-run.')
+            return
+
+        # Once all ingest rules have been applied, de-duplicate and then bulk insert
+        ingests = self._deduplicate_ingest_list(self.scan_id, ingests)
+
+        # bulk insert remaining as queued
+        with transaction.atomic():
+            Ingest.objects.bulk_create(ingests)
+
+        self._start_ingest_tasks(ingests)
+
+    @staticmethod
+    def _deduplicate_ingest_list(scan_id, new_ingests):
+        """Check the ingest records to ensure these ingests are not already created by previous scan run
+        
+        :param scan_id: ID of scan to check against
+        :type scan_id: integer
+        :param new_ingests: List of ingest models to validate for uniqueness
+        :type new_ingests: :class:`ingest.models.Ingest`
+        :returns: List of deduplicated ingest models
+        :rtype: List[:class:`ingest.models.Ingest`]
+        """
+
+        deduplicate_file_names = set()
+        list_count = len(new_ingests)
+        ingest_file_names = [ingest.file_name for ingest in new_ingests]
+
+        existing_ingests = Ingest.objects.get_ingests_by_scan(scan_id, ingest_file_names)
+        existing_ingest_file_names = [ingest.file_name for ingest in existing_ingests]
+
+        deduplicated_ingests = []
+        for ingest in new_ingests:
+            if ingest.file_name not in deduplicate_file_names:
+                deduplicate_file_names.add(ingest.file_name)
+                deduplicated_ingests.append(ingest)
+            else:
+                logging.info('Removed duplicate file_name %s from ingests at file_path %s',
+                             ingest.file_name, ingest.file_path)
+
+        final_ingests = [x for x in deduplicated_ingests if x.file_name not in existing_ingest_file_names]
+
+        logger.info('Removed %i duplicates of pre-existing ingests.', list_count - len(final_ingests))
+
+        return final_ingests
+
+    def _process_ingest(self, file_path, file_size):
```

Comment:
This seems like it should be generic to both scan and strike. There should be one code path for creating/setting up an ingest record and filling out the fields based on file ingest rules.